### PR TITLE
Add docs/TRANSLATION.md

### PR DESCRIPTION
**Pull request by dhilst**

### CODE REVIEW

File: docs/TRANSLATION.md [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+
+If  `[a]` is impure we need to use the monadic notation
+
+```
+  [let foo = a; ...]
+------------------------
+let** foo := [a] in [...]
+```
+* Note that `let*` will drive the type of `...`.
+* @TODO Can we know if `a` is impore or not?
```
Comment (clarus) [inline docs/TRANSLATION.md:145]: We can assume that all the functions are impure to start with.

File: docs/TRANSLATION.md [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+* @TODO `foo` is bounded in `...`. I had to take this into account
+  in coq-of-solidity because when I find the identifier `foo` inside
+  the `...` I need to know if it was local or not, to guide the translation.
+  Non local variables, i.e., state variables were translated to `state.(foo)`
+  while local variables stays like `foo`. In coq-of-solidity I use a map
+  as an environment to keep the scope of the variables (among other
+  things). I think that this is not a problem in rust because `self`
+  need to be explicited in order to be accessed.
```
Comment (clarus) [inline docs/TRANSLATION.md:153]: Yes indeed `self` is like a normal variable.

File: docs/TRANSLATION.md [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+[let Ctr foo = a else { ...0 ] }; ...1 ]
+----------------------------------------
+       let** foo := match [a] with 
+         | Ctr foo => [a] 
+         | _ => [ ...0 ]
+         end in
+         [ ...1 ]
+```
+
+* @TODO: Understand the `let**` notation, I'm just guessing that it
```
Comment (clarus) [inline docs/TRANSLATION.md:176]: I added the `let**` as an optimization of the `let*` for some cases, but I am not sure if we should keep it. The two should be equivalent.

File: docs/TRANSLATION.md [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+  * In coq-of-solidity I had to iterate over the expressions looking
+    for function calls, for other reason, but this can be used in the
+    translation to tell if an if expression is pure or not.
+
+## Function calls
+
+```
+  [f(a, b)]
+-------------
+ f(|[a] [b]|)
```
Comment (clarus) [inline docs/TRANSLATION.md:259]:
```suggestion
 f(|[a], [b]|)
```

File: docs/TRANSLATION.md (hunk 3 above)
Comment (dhilst) [reply docs/TRANSLATION.md:176]: Okay I will use `let*` for now

**Comment by dhilst** [comment]
Closing for now to focus on ERC20

**Pull request closed without merging**